pkg/chains/storage/grafeas/grafeas.go

```diff
@@ -186,117 +184,191 @@ func (b *Backend) Type() string {
 	noteID := b.cfg.Storage.Grafeas.NoteID
 
-	b.logger.Infof("Creating a note with note name %s", noteID)
-
-	// create note request
-	noteReq := &pb.CreateNoteRequest{
-		Parent: b.getProjectPath(),
-		NoteId: noteID,
-		Note: &pb.Note{
-			ShortDescription: "An attestation note",
-			Kind:             commonpb.NoteKind_ATTESTATION,
-			Type: &pb.Note_AttestationAuthority{
-				AttestationAuthority: &attestationpb.Authority{
-					Hint: &attestationpb.Authority_Hint{
-						HumanReadableName: "This note was auto-generated by Tekton Chains",
+	// for oci image: AttestationNote
+	if opts.PayloadFormat == formats.PayloadTypeSimpleSigning {
+		return b.client.CreateNote(ctx,
+			&pb.CreateNoteRequest{
+				Parent: b.getProjectPath(),
+				NoteId: fmt.Sprintf(attestationNoteNameFormat, noteID),
+				Note: &pb.Note{
+					ShortDescription: "OCI Artifact Attestation Note",
+					Type: &pb.Note_Attestation{
+						Attestation: &pb.AttestationNote{
+							Hint: &pb.AttestationNote_Hint{
+								HumanReadableName: "This attestation note was generated by Tekton Chains",
+							},
+						},
 					},
 				},
 			},
-		},
+		)
 	}
 
-	// store note
-	if _, err := b.client.CreateNote(ctx, noteReq); err != nil {
-		return err
+	// for taskrun: BuildNote
+	return b.client.CreateNote(ctx,
+		&pb.CreateNoteRequest{
+			Parent: b.getProjectPath(),
+			NoteId: fmt.Sprintf(buildNoteNameFormat, noteID),
+			Note: &pb.Note{
+				ShortDescription: "Build Provenance Note",
+				Type: &pb.Note_Build{
+					Build: &pb.BuildNote{
+						BuilderVersion: tr.GetGroupVersionKind().GroupVersion().String(),
+					},
+				},
+			},
+		},
+	)
+}
+
+// createOccurrence creates grafeas occurrences in the grafeas server that stores the original payload and its signature
+// for a single oci artifact
+//   - its simplesigning payload and signature will be stored in ATTESTATION occurrence
+//   - the identifier/ResourceUri is IMAGE_URL@IMAGE_DIGEST
+//
+// for a taskrun object
+//   - its intoto payload and signature will be stored in a BUILD occurrences for each image artifact generated from the taskrun
+//   - each BUILD occurrence will have the same data but differ in the ResourceUri field
+//   - the identifier/ResourceUri is https://IMAGE_URL@IMAGE_DIGEST
+func (b *Backend) createOccurrence(ctx context.Context, tr *v1beta1.TaskRun, payload []byte, signature string, opts config.StorageOpts) ([]*pb.Occurrence, error) {
+	occs := []*pb.Occurrence{}
+
+	uri, allURIs := b.retrieveOCIURIs(tr, opts)
+
+	// create Occurrence_Attestation for OCI
+	if opts.PayloadFormat == formats.PayloadTypeSimpleSigning {
+		occ, err := b.createAttestationOccurrence(ctx, tr, payload, signature, uri)
+		if err != nil {
+			return nil, err
+		}
+		occs = append(occs, occ)
+		return occs, nil
 	}
 
-	return nil
+	// create Occurrence_Build for TaskRun
+	for _, uri := range allURIs {
+		occ, err := b.createBuildOccurrence(ctx, tr, payload, signature, fmt.Sprint("https://", uri))
```

Comment:
Not sure if we should do this 🤔 

What kind of URIs are we expecting here, and why do we need to append the https (e.g. if it's needed, why isn't it already present)?

Reply:
> What kind of URIs are we expecting here?

URIs are the identifiers of the container images built from the taskrun (`image_url@image_digest`).

> Not sure if we should do this 🤔

The reason why we create build occurrence per image built with identical intoto payload and signature but differing in the `resourceUri` field is that we want to enable users to retrieve provenance for a particular artifact built. 

For example, if a taskrun produces two images identified by `foo` and `bar`, we want to have the provenance associated with each of them because it's the same taskrun that produces them. This would allow users/provenance consumers to retrieve the provenance for a specific artifact by filtering the `resourceUri` field that is set to be the image identifier.

> why do we need to append the https (e.g. if it's needed, why isn't it already present)?

For attestation occurrence that stores simplesigning attestation for OCI artifact, the resourceUri is just set to be `image_url@image_digest`. 
For build occurrence that contains intoto attestation for the taskrun, we want to add a prefix `https://` to differentiate the build occurrence resourceUri and attestation occurrence resourceUri.

Example: taskrun `X` generates a single image artifact `URL@Digest`. 
- If users want to see the image simplesigning attestation, they can use `URL@Digest` as the filter of [ListOccurrences](https://cloud.google.com/container-analysis/docs/reference/rest/v1/projects.occurrences/list)
- If users want to see the provenance (how the image was built), they can use `https://URL@Digest` as the filter to retrieve the taskrun `X`'s intoto provenance.

> why isn't it already present?
Because prior to this, we use [K8s resource uri](https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-urist) as the resourceUri for taskrun payload. And that doesn't work for consumers i.e. BinAuthz to look up occurrences because they only know the identifier of the deployed artifact at deployment time.


These are great questions. Please let me know if that makes sense. Feel free to follow up if you have more questions.

Reply:
Discussed offline with @sherzberg-1 about this. Thanks @sherzberg-1 for the help and sharing the knowledge!

Just realized that here we can also use `IMAGE_URL@IMAGE_DIGEST` for the build occurrence without having `https://` as the prefix. In other words, if a taskrun generates an image called `foo@bar`, we create one attestation occurrence with `foo@bar` as the resourceUri field, and one build occurrence with `foo@bar` as the resourceUri field too. For consumers, they can filter on both the `resourceUri` and the `kind` field to look up build OR attestation occurrence based on their needs. 
nit: two occurrences can have same resourceUri as long as they are linked to different grafeas notes.

Let me know what you think. @wlynch Thanks!

Reply:
Cool, let's use the original identifier then then so we're not mucking with any inputs the user gives us.

Reply:
Done!